schema/pgsql/schema.sql

```diff
@@ -278,3 +278,32 @@ CREATE TABLE incident_history (
     CONSTRAINT pk_incident_history PRIMARY KEY (id),
     FOREIGN KEY (incident_id, rule_escalation_id) REFERENCES incident_rule_escalation_state(incident_id, rule_escalation_id)
 );
+
+CREATE OR REPLACE FUNCTION now_extended() RETURNS bigint
+    LANGUAGE SQL AS
+$$
+SELECT ((EXTRACT(EPOCH FROM CURRENT_TIMESTAMP) * 1_000) :: bigint);
+$$;
+
+CREATE TABLE session
+(
+    id               VARCHAR(256) NOT NULL,
+    username         VARCHAR(254) NOT NULL,
+    device_id        VARCHAR(8)   NOT NULL,
+    authenticated_at bigint       NOT NULL DEFAULT now_extended(),
+
+    CONSTRAINT pk_id_username_device PRIMARY KEY (id, username, device_id)
+);
+
+CREATE INDEX idx_id ON session (
+                                id
+    );
```

Comment:
I do not agree on that one. Defining a primary key which consists of multiple columns only creates an index for searches that do provide all of those fields.
As we need to check for the existence of a session identifier when hooking into the authentication mechanism to mitigate the reallocation problem stated above, we want a fast lookup for the session id only.
This might be marginal and not really important even in big installations with a lot of concurrent users, but I tried to respect the scalability as much as I could.

Reply:
> Defining a primary key that consists of multiple columns only creates an index for lookups that do provide all of those fields. This is at least the case for PostgreSQL.

I'd find that very surprising and also, the [documentation](https://www.postgresql.org/docs/16/indexes-multicolumn.html) says something different:

> A multicolumn B-tree index can be used with query conditions that involve any subset of the index's columns, but the index is most efficient when there are constraints on the leading (leftmost) columns.

Reply:
> I'd find that very surprising and also, the [documentation](https://www.postgresql.org/docs/16/indexes-multicolumn.html) says something different:
> 
> > A multicolumn B-tree index can be used with query conditions that involve any subset of the index's columns, but the index is most efficient when there are constraints on the leading (leftmost) columns.

I seem to have missed that specific phrase while reading through the index definitions. That was my bad and the additional single-dimension index is not needed then. Going to remove it :)